### ISSUE TITLE
2020-02-11 M4 AET values not generated for all columns when using actual time

### ISSUE THREAD

**Issue by jhhughes256** [tensfeldt/openNCA#178]
### Issue

When running M4 methods, there are testcases where AET is not generated for specific columns or SDEIDs. tc008 M4SS is used as an example below where we can see that `SDEID == 225930823` does not have any values for `AET` generated in any column. This is despite valid `PKCNCN` and `PKSMVL` values being available in the original data for this individual.

This occurs for all M4 testcases when using actual time (tc008, tc017, tc018, tc105, tc111, tc116). This issue has been separated from #177 for clarity.

```r
> r[c("SDEID", "AMT.24.08", "AMT.24.13", "AE.24.08", "AE.24.13")]
      SDEID AMT.24.08 AMT.24.13 AE.24.08  AE.24.13
1 225930823        NA  12.73858       NA        NA
2 363476907 12.213348        NA       NA 12.213348
3 532644281  9.807498        NA       NA  9.807498
4 754426795 16.942880        NA       NA 16.942880
5 926177600 16.375412        NA       NA 16.375412
6 927450954 32.388760        NA       NA 32.388760
> dplyr::arrange(d[c(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)], SDEID)
      SDEID PKATPD PKATPDE PKCNCN  PKSMVL PKCNCU PKSMVLU
1 225930823      0  24.133   9230 1407.73  NG/ML      GM
2 363476907      0  24.083   9970 1249.51  NG/ML      GM
3 532644281      0  24.083   8280 1208.17  NG/ML      GM
4 754426795      0  24.083   6170 2800.93  NG/ML      GM
5 926177600      0  24.083   6450 2589.60  NG/ML      GM
6 927450954      0  24.083   9310 3548.50  NG/ML      GM
```

Note that tc017 has nominal and actual times are the same, but the only the actual time run doesn't generate output for `AE.48.00`.

```r
> # tc017 with nominal time
> mct$TIME
[1] "NOMINAL"
> r
      SDEID DOSE1    AMT.4.00 AMT.12.00 AMT.24.00 AMT.48.00     AE.4.00  AE.12.00  AE.24.00   AE.48.00 AEPCT.4.00 AEPCT.12.00
1  58098661   200  1.62334324 56.971298  8.057586  4.150139  1.62334324 58.594641 66.652228  70.802366  0.8116716    29.29732
2  58613088   300 41.92012941 35.572466 11.807324 18.040106 41.92012941 77.492595 89.299919 107.340024 13.9733765    25.83087
3 107790547   300 25.05333333 41.907250 13.426541  7.244447 25.05333333 66.960583 80.387125  87.631572  8.3511111    22.32019
> # tc017 with actual time
> mct$TIME
[1] "ACTUAL"
> r
      SDEID DOSE1    AMT.4.00 AMT.12.00 AMT.24.00 AMT.48.00     AE.4.00  AE.12.00  AE.24.00 AE.48.00 AEPCT.4.00 AEPCT.12.00
1  58098661   200  1.62334324 56.971298  8.057586  4.150139  1.62334324 58.594641 66.652228       NA  0.8116716    29.29732
2  58613088   300 41.92012941 35.572466 11.807324 18.040106 41.92012941 77.492595 89.299919       NA 13.9733765    25.83087
3 107790547   300 25.05333333 41.907250 13.426541  7.244447 25.05333333 66.960583 80.387125       NA  8.3511111    22.32019
> dplyr::arrange(d[c(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)], SDEID)
       SDEID PKATPD PKATPDE  PKCNCN   PKSMVL PKCNCU PKSMVLU
1   58098661  0.000       4  3990.0  414.990  NG/ML      GM
2   58098661  4.000      12 62800.0  925.330  NG/ML      GM
3   58098661 12.000      24  8350.0  984.280  NG/ML      GM
4   58098661 24.000      48  4540.0  932.410  NG/ML      GM
5   58613088  0.000       4 66600.0  642.020  NG/ML      GM
6   58613088  4.000      12 44300.0  819.050  NG/ML      GM
7   58613088 12.000      24 31400.0  383.550  NG/ML      GM
8   58613088 24.000      48  9630.0 1910.790  NG/ML      GM
9  107790547  0.000       4 27200.0  939.500  NG/ML      GM
10 107790547  4.000      12 35700.0 1197.350  NG/ML      GM
11 107790547 12.000      24  8190.0 1672.170  NG/ML      GM
12 107790547 24.000      48  2210.0 3343.591  NG/ML      GM
```

**Comment by jhhughes256** [comment]
### Additional Findings

When running tc018 M4SS `AET` values are generated for only one column (similar to tc008 M4SD and M4SS). When looking at the `PKATPDE` value for this individual, neither value is equal to 10.00.

```r
> # tc018 M4SS with actual time
> # determine which columns have non-NA values
> r[stringr::str_detect(names(r), "^AE.")] %>% 
+   dplyr::summarise_all(dplyr::funs(sum(!is.na(.)) > 0)) %>% 
+   dplyr::select_if(rlang::is_true)
  AE.10.00
1     TRUE
> # determine which SDEID has non-NA values for that column and observe original times in dataset
> dplyr::filter(r, !is.na(AE.10.00)) %>%
+   dplyr::pull(SDEID) %>%
+   {dplyr::filter(d, SDEID == .)} %>%
+   dplyr::select(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)
      SDEID PKATPD PKATPDE PKCNCN PKSMVL PKCNCU PKSMVLU
1 303069272  1.067   4.167   5520    780  NG/ML      ML
2 303069272 12.033  23.033   2230   1380  NG/ML      ML
```

When running tc105 M4SS the AE columns match the time from the data, however only one column has values generated.

```r
> # tc105 M4SS with actual time output
> r[c("SDEID", "AE.24.08", "AE.24.13")]
       SDEID  AE.24.08 AE.24.13
1  661083556 32.388760       NA
2  737250770        NA       NA
3 1723871653  9.807498       NA
4 2187657022 12.213348       NA
5 2432717605 16.942880       NA
6 3854759896 16.375412       NA
> # tc105 M4SS with actual time input dataset
> dplyr::arrange(d[c(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)], SDEID)
       SDEID PKATPD  PKATPDE PKCNCN  PKSMVL PKCNCU PKSMVLU
1  661083556      0 24.08333   9310 3548.50  NG/ML      GM
2  737250770      0 24.13333   9230 1407.73  NG/ML      GM
3 1723871653      0 24.08333   8280 1208.17  NG/ML      GM
4 2187657022      0 24.08333   9970 1249.51  NG/ML      GM
5 2432717605      0 24.08333   6170 2800.93  NG/ML      GM
6 3854759896      0 24.08333   6450 2589.60  NG/ML      GM
```

When running tc111 M4SD there are four columns which have non-NA values (`AE.0.00`, `AE.11.05`, `AE.11.07`, `AE.11.08`), however they have values for three SDEIDs. They all have times that match with `AE.0.00` but don't have values that match for the other three columns.

```r
> # tc111 M4SD with actual time
> # determine which columns have non-NA values (picks up AEPCT too, coz regex)
> r[stringr::str_detect(names(r), "^AE.")] %>% 
+    dplyr::summarise_all(dplyr::funs(sum(!is.na(.)) > 0)) %>% 
+    dplyr::select_if(rlang::is_true)
  AE.0.00 AE.11.05 AE.11.07 AE.11.08 AEPCT.0.00 AEPCT.11.05 AEPCT.11.07 AEPCT.11.08 AEPCT
1    TRUE     TRUE     TRUE     TRUE       TRUE        TRUE        TRUE        TRUE  TRUE
> # show which rows in the output have non-NA values
>  r[stringr::str_detect(names(r), "^AE.")] %>% 
+   dplyr::summarise_all(dplyr::funs(sum(!is.na(.)) > 0)) %>%     dplyr::select_if(rlang::is_true) %>%
+   names %>% 
+   {dplyr::filter(r, !is.na(!!as.name(.)))}
       SDEID       DOSE   DOSEC AE.0.00 AE.11.05 AE.11.07 AE.11.08 AE.12.03 AE.12.05 AE.12.07 AE.23.05 AE.24.05 AE.24.07 AE.24.08
1 2579639686 1093027000 1093027       0        0        0        0       NA       NA       NA       NA       NA       NA       NA
2 3021231022 1099202000 1099202       0        0        0        0       NA       NA       NA       NA       NA       NA       NA
3 4134187385 1100586000 1100586       0        0        0        0       NA       NA       NA       NA       NA       NA       NA
  AE.47.05 AE.47.07 AE.48.05 AE.48.07 AE.48.08 AE.48.10 AE.48.15 AE.71.05 AE.71.07 AE.71.08 AE.72.03 AE.72.05 AE.72.07 AE.72.08
1       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
2       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
3       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
  AE.95.03 AE.95.05 AE.95.07 AE.96.07 AE.96.08 AE.96.17 AE.119.02 AE.119.03 AE.120.02 AE.143.05 AE.143.07 AE.143.10 AE.144.00
1       NA       NA       NA       NA       NA       NA        NA        NA        NA        NA        NA        NA        NA
2       NA       NA       NA       NA       NA       NA        NA        NA        NA        NA        NA        NA        NA
3       NA       NA       NA       NA       NA       NA        NA        NA        NA        NA        NA        NA        NA
  AE.144.02 AE.144.03 AE.144.10 AE.168.02 AE.168.05 AE.168.18 AE.192.02 AE.192.10 AE.216.07 AE.240.02 AEPCT.0.00 AEPCT.11.05
1        NA        NA        NA        NA        NA        NA        NA        NA        NA        NA          0           0
2        NA        NA        NA        NA        NA        NA        NA        NA        NA        NA          0           0
3        NA        NA        NA        NA        NA        NA        NA        NA        NA        NA          0           0
> # determine which SDEIDs have non-NA values for those columns and observe original times in dataset
> r[stringr::str_detect(names(r), "^AE.")] %>% 
+    dplyr::summarise_all(dplyr::funs(sum(!is.na(.)) > 0)) %>% 
+    dplyr::select_if(rlang::is_true) %>%
+    names %>% 
+    {dplyr::filter(r, !is.na(!!as.name(.)))} %>%
+    dplyr::pull(SDEID) %>%
+    {dplyr::filter(d, SDEID %in% .)} %>%
+    dplyr::select(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)
        SDEID PKATPD_ORIG PKATPDE_ORIG    PKCNCN PKSMVL PKCNCU PKSMVLU
1  3021231022       0.000        0.000    0.0000  138.4 DPM/ML      ML
2  3021231022       0.000       12.067  732.0000 1220.6 DPM/ML      ML
3  3021231022      12.067       24.067   49.4000  660.2 DPM/ML      ML
4  3021231022      24.067       48.083    9.5200 1206.6 DPM/ML      ML
5  3021231022      48.083       72.083    0.8160 2123.6 DPM/ML      ML
6  3021231022      72.083       96.067    0.0402 2186.3 DPM/ML      ML
7  3021231022      96.067      120.017    0.0000 1840.7 DPM/ML      ML
8  3021231022     120.017      144.033    0.0000 2674.3 DPM/ML      ML
9  3021231022     144.033      168.017    0.0000 2135.6 DPM/ML      ML
10 3021231022     168.017      192.017    0.0000 2091.5 DPM/ML      ML
11 4134187385       0.000        0.000    0.0000  251.0 DPM/ML      ML
12 4134187385       0.000       12.067 1232.0000  708.2 DPM/ML      ML
13 4134187385      12.067       24.083   42.1000  512.8 DPM/ML      ML
14 4134187385      24.083       48.100    5.8900 1535.1 DPM/ML      ML
15 4134187385      48.100       72.050    1.7900 2130.2 DPM/ML      ML
16 4134187385      72.050       96.067    0.7970 1931.7 DPM/ML      ML
17 4134187385      96.067      120.017    0.3600 1628.0 DPM/ML      ML
18 4134187385     120.017      144.100    0.0367 2629.4 DPM/ML      ML
19 4134187385     144.100      168.017    0.0000 2581.9 DPM/ML      ML
20 4134187385     168.017      192.017    0.0000 1836.1 DPM/ML      ML
21 2579639686       0.000        0.000    0.0000  160.4 DPM/ML      ML
22 2579639686       0.000       12.050 2663.0000  376.5 DPM/ML      ML
23 2579639686      12.050       24.067   75.1000  296.5 DPM/ML      ML
24 2579639686      24.067       48.067    3.5500  783.5 DPM/ML      ML
25 2579639686      48.067       72.067    0.3870 1332.2 DPM/ML      ML
26 2579639686      72.067       96.167    0.0916  963.1 DPM/ML      ML
27 2579639686      96.167      120.017    0.0933  649.6 DPM/ML      ML
28 2579639686     120.017      144.017    0.0000  984.0 DPM/ML      ML
29 2579639686     144.017      168.183    0.0000  791.0 DPM/ML      ML
30 2579639686     168.183      192.100    0.0000 1225.2 DPM/ML      ML
```

When running tc116 M4SD, values are generated for both columns regardless of whether that SDEID has that time as a `PKATPDE` value. Additionally they are missing for the one SDEID with a unique `PKATPDE` value, for what seems like no reason as there is both a valid `PKCNCN` and `PKSMVL` value available for calculation of `AET`.

```r
> # tc116 M4SS with actual time output data
> r[c("SDEID", "AE.24.08", "AE.24.13")]
       SDEID  AE.24.08  AE.24.13
1  661083556 32.388760 32.388760
2  737250770        NA        NA
3 1723871653  9.807498  9.807498
4 2187657022 12.213348 12.213348
5 2432717605 16.942880 16.942880
6 3854759896 16.375412 16.375412
> # tc116 M4SS with actual time input dataset
> dplyr::arrange(d[c(mct$SDEID, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$AMOUNT, mct$CONCU, mct$AMOUNTU)], SDEID)
       SDEID     PKATPD  PKATPDE PKCNCN  PKSMVL PKCNCU PKSMVLU
1  661083556 -0.1166667 24.08333   9310 3548.50  NG/ML      GM
2  737250770 -0.1166667 24.13333   9230 1407.73  NG/ML      GM
3 1723871653 -0.1666667 24.08333   8280 1208.17  NG/ML      GM
4 2187657022 -0.1666667 24.08333   9970 1249.51  NG/ML      GM
5 2432717605 -0.1000000 24.08333   6170 2800.93  NG/ML      GM
6 3854759896 -0.1666667 24.08333   6450 2589.60  NG/ML      GM
```

**Comment by jhhughes256** [comment]
As of commit 4097880 this issue is partially blocked by #182 and #183. tc017 M4SD is not affect by this error however no change in behaviour has been observed.

**Comment by jhhughes256** [comment]
As of commit 231f2da this is mostly fixed. I've found some differences in ACTENDTIMEs and AE columns for tc111 M4SD which I've shown below. Notice that there is a value under PKATPDE_ORIG for time 12.05, but no value in the corresponding column. For other individuals 0 is NA, which I'm not sure if that is desired behaviour.

All other testcases work though, just tc111 which is behaving like this.

```r
        SDEID PKATPD_ORIG PKATPDE_ORIG    PKCNCN PKSMVL PKCNCU PKSMVLU
21 2579639686       0.000        0.000    0.0000  160.4 DPM/ML      ML
22 2579639686       0.000       12.050 2663.0000  376.5 DPM/ML      ML
23 2579639686      12.050       24.067   75.1000  296.5 DPM/ML      ML
24 2579639686      24.067       48.067    3.5500  783.5 DPM/ML      ML
25 2579639686      48.067       72.067    0.3870 1332.2 DPM/ML      ML
26 2579639686      72.067       96.167    0.0916  963.1 DPM/ML      ML
27 2579639686      96.167      120.017    0.0933  649.6 DPM/ML      ML
28 2579639686     120.017      144.017    0.0000  984.0 DPM/ML      ML
29 2579639686     144.017      168.183    0.0000  791.0 DPM/ML      ML
30 2579639686     168.183      192.100    0.0000 1225.2 DPM/ML      ML
>  dplyr::filter(r, SDEID == 2579639686)
       SDEID       DOSE   DOSEC AE.0.00 AE.11.05 AE.11.07 AE.11.08 AE.12.03 AE.12.05 AE.12.07
1 2579639686 1093027000 1093027      NA       NA       NA       NA       NA       NA       NA
  AE.23.05 AE.24.05 AE.24.07 AE.24.08 AE.47.05 AE.47.07 AE.48.05 AE.48.07 AE.48.08 AE.48.10 AE.48.15
1       NA       NA 22267150       NA       NA       NA       NA 25048575       NA       NA       NA
  AE.71.05 AE.71.07 AE.71.08 AE.72.03 AE.72.05 AE.72.07 AE.72.08 AE.95.03 AE.95.05 AE.95.07 AE.96.07
1       NA       NA       NA       NA       NA 25564136       NA       NA       NA       NA       NA
  AE.96.08 AE.96.17 AE.119.02 AE.119.03 AE.120.02 AE.143.05 AE.143.07 AE.143.10 AE.144.00 AE.144.02
1       NA 25652356        NA        NA  25712964        NA        NA        NA        NA  25712964
  AE.144.03 AE.144.10 AE.168.02 AE.168.05 AE.168.18 AE.192.02 AE.192.10 AE.216.07 AE.240.02
1        NA        NA        NA        NA  25712964        NA  25712964        NA        NA
```

This occurs for the following SDEID for some value of mct$ACTENDTIME

```r
        SDEID                             data
1   469690710                             TRUE
2   768248218                             TRUE
3  2210432242  Numeric: lengths (9, 10) differ
4  2579639686  Numeric: lengths (8, 10) differ
5  2963738954                             TRUE
6  3021231022  Numeric: lengths (8, 10) differ
7  3194419488                             TRUE
8  3589284518  Numeric: lengths (9, 10) differ
9  4020796352 Numeric: lengths (11, 12) differ
10 4134187385  Numeric: lengths (8, 10) differ
11 4217055814                             TRUE
```

Code I've used to find these differences

```r
# worked for tc111
r %>%
  dplyr::mutate(SDEID2 = SDEID) %>%
  dplyr::group_by(SDEID) %>% tidyr::nest() %>%
  dplyr::mutate(data = purrr::map(data, function(df) {
    aecols <- names(df)[stringr::str_detect(names(df), "^AE(?:.+)\\d\\d$")]
    aecols <- aecols[!stringr::str_detect(aecols, "^AEPCT")]
    colnums <- dplyr::select_at(df, aecols) %>%
      dplyr::select_if({!is.na(.)}) %>%
      names %>%
      stringr::str_replace("AE.", "") %>%
      parse_double
    endtimes <- dplyr::arrange(d, SDEID, rlang::UQ(as.name(mct$ACTTIME)), rlang::UQ(as.name(mct$ACTENDTIME))) %>%
      dplyr::filter(SDEID == df$SDEID2) %>%
      dplyr::pull(mct$ACTENDTIME) %>%
      round(2)
    as.character(all.equal(colnums, endtimes))
  })) %>%
  tidyr::unnest() %>%
  as.data.frame()
```

```r
> # tc018 M4SS with actual time
> r[stringr::str_detect(names(r), "^AE.")] %>% 
+    dplyr::summarise_all(dplyr::funs(sum(!is.na(.)) > 0)) %>% 
+    dplyr::select_if(rlang::is_true)
  AE.4.17 AE.6.03 AE.6.52 AE.7.28 AE.9.50 AE.9.60 AE.9.75 AE.9.90 AE.9.92 AE.10.00 AE.10.05 AE.10.10
1    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE    TRUE     TRUE     TRUE     TRUE
  AE.10.38 AE.10.50 AE.10.57 AE.10.73 AE.10.87 AE.10.92 AE.11.38 AE.11.43 AE.11.48 AE.11.55 AE.11.77
1     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE
  AE.11.85 AE.11.87 AE.11.88 AE.11.90 AE.11.93 AE.12.00 AE.15.90 AE.20.88 AE.21.98 AE.22.57 AE.22.58
1     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE
  AE.22.60 AE.22.63 AE.22.65 AE.22.80 AE.22.82 AE.22.87 AE.23.03 AE.23.12 AE.23.20 AE.23.35 AE.23.42
1     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE
  AE.23.75 AE.23.77 AE.23.85 AE.23.88 AE.23.92 AE.24.28
1     TRUE     TRUE     TRUE     TRUE     TRUE     TRUE
> r %>%
+   dplyr::mutate(SDEID2 = SDEID) %>%
+   dplyr::group_by(SDEID) %>% tidyr::nest() %>%
+   dplyr::mutate(data = purrr::map(data, function(df) {
+     aecols <- names(df)[stringr::str_detect(names(df), "^AE(?:.+)\\d\\d$")]
+     colnums <- dplyr::select_at(df, aecols) %>%
+       dplyr::select_if({!is.na(.)}) %>%
+       names %>%
+       stringr::str_replace("AE.", "") %>%
+       parse_double
+     endtimes <- dplyr::filter(d, SDEID == df$SDEID2) %>%
+       dplyr::pull(mct$ACTENDTIME) %>%
+       round(2)
+     all.equal(colnums, endtimes)
+   })) %>%
+   tidyr::unnest() %>%
+   as.data.frame()
       SDEID data
1   43456658 TRUE
2   98862581 TRUE
3  102652097 TRUE
4  155315252 TRUE
5  212128320 TRUE
6  232007206 TRUE
7  303069272 TRUE
8  322184155 TRUE
9  346821548 TRUE
10 359622627 TRUE
11 392430764 TRUE
12 395827329 TRUE
13 436516214 TRUE
14 446296844 TRUE
15 448488588 TRUE
16 499736026 TRUE
17 514418530 TRUE
18 522999139 TRUE
19 553304375 TRUE
20 578457836 TRUE
21 581080171 TRUE
22 582170099 TRUE
23 583844675 TRUE
24 590276539 TRUE
25 598462265 TRUE
26 658062633 TRUE
27 678253176 TRUE
28 702151038 TRUE
29 722709781 TRUE
30 724829716 TRUE
31 763781430 TRUE
32 774566237 TRUE
33 813912334 TRUE
34 846570468 TRUE
35 850772340 TRUE
36 869199587 TRUE
37 882441025 TRUE
38 889990267 TRUE
39 904183466 TRUE
40 909697662 TRUE
41 927550096 TRUE
42 934267135 TRUE
43 936232252 TRUE
44 974609637 TRUE
45 995384974 TRUE

**Comment by jhhughes256** [comment]
As of commit 626e458 this issue remains resolved for the following testcases: tc008, tc017, tc105, tc116. 

This issue still affects tc018 M4SS and tc111 M4SD. The output below shows the `SDEID` columns where the `AET` times from `data_out` do not match up with the `mct$ACTENDTIME` column from the input data.

```r
> # tc018
      SDEID data                          
      <int> <chr>                         
1  43456658 Numeric: lengths (1, 2) differ
2 102652097 Numeric: lengths (1, 2) differ
3 346821548 Numeric: lengths (0, 2) differ
4 359622627 Numeric: lengths (1, 2) differ
5 436516214 Numeric: lengths (1, 2) differ
6 583844675 Numeric: lengths (1, 2) differ
7 678253176 Numeric: lengths (1, 2) differ
8 774566237 Numeric: lengths (1, 2) differ
9 850772340 Numeric: lengths (1, 2) differ
> # tc111
        SDEID                             data
1   469690710   Numeric: lengths (6, 7) differ
2  2210432242  Numeric: lengths (9, 10) differ
3  2579639686  Numeric: lengths (8, 10) differ
4  2963738954   Numeric: lengths (6, 7) differ
5  3021231022  Numeric: lengths (9, 10) differ
6  3589284518  Numeric: lengths (9, 10) differ
7  4020796352 Numeric: lengths (11, 12) differ
8 4134187385  Numeric: lengths (9, 10) differ
9 4217055814   Numeric: lengths (6, 7) differ
```

Looking at `AE.0.00` we can see that it is `NA` for four `SDEID`s when there should be a value `SDEID %in% c(2210432242, 2579639686, 3589284518, 4020796352)`. Excluding those (as `AE.0.00` has been split into a separate issue) there are five `SDEID`s with one `AET` value not generated `SDEID %in% c(469690710, 2963738954, 3021231022, 4134187385, 4217055814)`. 

For these cases the times at which the values are missing don't seem to have a pattern. An example of the output for `SDEID == 469690710` is provided in the snippet below, using the `checkaet_fn` function I made to help with 

```r
checkaet_fn <- function(id, indata = d, outdata = r, map = mct) {
  list(
    indata = indata[indata$SDEID == id, c(map$SDEID, map$ACTTIME, map$ACTENDTIME, map$CONC, map$AMOUNT)],
    outdata = outdata[outdata$SDEID == id, stringr::str_detect(names(r), "^AE\\.(?:.+)\\d\\d$")]
  )
}
> # SDEID == 469690710
> checkaet_fn(43456658)
$indata
      SDEID PKATPD PKATPDE PKCNCN PKSMVL
11 43456658  1.350   6.517  11300    560
12 43456658 12.033  20.883   5720    650

$outdata
  AE.4.17 AE.6.03  AE.6.52 AE.7.28 AE.9.50 AE.9.60 AE.9.75 AE.9.90 AE.9.92 AE.10.00 AE.10.05
1      NA      NA 10046000      NA      NA      NA      NA      NA      NA       NA       NA
  AE.10.10 AE.10.38 AE.10.50 AE.10.57 AE.10.73 AE.10.87 AE.10.92 AE.11.38 AE.11.43 AE.11.48 AE.11.55
1       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
  AE.11.77 AE.11.85 AE.11.87 AE.11.88 AE.11.90 AE.11.93 AE.12.00 AE.15.90 AE.20.88 AE.21.98 AE.22.57
1       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
  AE.22.58 AE.22.60 AE.22.63 AE.22.65 AE.22.80 AE.22.82 AE.22.87 AE.23.03 AE.23.12 AE.23.20 AE.23.35
1       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA       NA
  AE.23.42 AE.23.75 AE.23.77 AE.23.85 AE.23.88 AE.23.92 AE.24.28
1       NA       NA       NA       NA       NA       NA       NA
```

**Comment by rudraya** [comment]
After further investigation for this issue, I believe the discrepancy is generated because of the logic that was put into place with respect to:
- Removing duplicate TIME and CONC values for a given SDEID
- Removing duplicate TIME but different CONC for a given SDEID

Based on that this is the warnings that I get when I run tc111 with TIME == "ACTUAL, "ACTTIME" == "PKATPD_ORIG" and "ACTENDTIME == "PKATPDE_ORIG"
```r
> warnings()
Warning messages:
1: In validate_timeconc_data(map, data, flag, verbose = verbose) :
  No tau information provided in 'map'.
2: In validate_timeconc_data(map, data, flag, verbose = verbose) :
  No told information provided in 'map'.
3: In openNCA::run_computation(data = data_tc111, map = map_tc111,  ... :
  'TAUs and TOLDs values are not provided via 'flag'! Using values provided via 'data' instead!
4: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'FLGACCEPTKELCRIT' values provided via 'map' does not have a valid parameter name 'KELRSQ'
5: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'FLGACCEPTKELCRIT' values provided via 'map' does not have a valid parameter name 'KELNOPT'
6: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'FLGEMESIS' is not present in the dataset
7: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'FLGACCEPTPREDOSECRIT' cannot be computed if 'CMAX' is not part of the calculated parameters
8: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'SPANRATIOCRIT' is not present in the dataset
9: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
10: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
11: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '2210432242' due to duplicate TIME but different CONC values
12: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
13: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '2579639686' due to duplicate TIME but different CONC values
14: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
15: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
16: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '3021231022' due to duplicate TIME but different CONC values
17: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
18: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
19: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '3589284518' due to duplicate TIME but different CONC values
20: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
21: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '4020796352' due to duplicate TIME but different CONC values
22: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
23: In doTryCatch(return(expr), name, parentenv, handler) :
  Removing SDEID: '4134187385' due to duplicate TIME but different CONC values
24: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
25: In unit_conversion(data = data_data, map = map_data, result = data_data,  ... :
  'DOSEU' and/or 'DOSEOUTPUTUNIT' value provided via 'map' is not valid for unit conversion
26: In run_M4_SD_computation(data = merged_data, map = map_data,  ... :
  Flag 'FLGACCEPTKELCRIT' values provided via 'map' does not have a parameter name that is generated as an output 'AURCXPCTP'
27: In unit_conversion(data = data_data, map = map_data, result = computation_df,  ... :
  'PKATPDU' #7# is not present in the dataset provided via 'map'
28: In unit_conversion(data = data_data, map = map_data, result = computation_df,  ... :
  'PKATPDU', 'CONCU' and 'DOSEU' #8# are not present in the dataset provided via 'map'
29: In unit_conversion(data = data_data, map = map_data, result = computation_df,  ... :
  'CONCU' #24# is not present in the dataset provided via 'map'
30: In unit_conversion(data = data_data, map = map_data, result = computation_df,  ... :
  'CONCU', 'PKATPDU and 'NORMBSU' #19# are not present in the dataset provided via 'map'
```
This indicates issues with SDIED: 2210432242, 2579639686, 3021231022, 3589284518, 4020796352, 4134187385

Based on that this is the warnings that I get when I run tc018 with TIME == "ACTUAL, "ACTTIME" == "PKATPD" and "ACTENDTIME == "PKATPDE", I am seeing values for the SDEID "43456658" at time "20.883". I will push the code, so as of commit 8c30885, this shouldn't be an issue.

**Comment by tensfeldt** [comment]
We'll retest with this commit but also have to carefully test the duplicate TIME and TIME/CONC datapoint and FLGTIME replacement functionality. This will be done today.

**Comment by tensfeldt** [comment]
Initial retests are below with tc018 and commit 8c30885

The primary issue with the results with this test cases are that the AEs reported with the label including the sampling time interval are variously reported on one or the other of the sample times, not routinely on the last time. See below that for SDEID==4346658, AE of 10.046 is reported also on the start of the interval timepoint of AE.6.52 which is the start of the interval not the end of 20.88.

For SDEID==102652097, AE of 107.114 is reported also on the end of the interval timepoint  AE.23.75.

There doesn't appear to be a pattern associated with this. This value should always be reported on the end of the interval timepoint and the other timepoint should not be reported (unless other SDEIDs have that as a end of an interval)

```r
> sdeid <- 43456658 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE1, mct$DOSE1U)]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML") * 10^-6
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD PKATPDE PKCNCN PKCNCU PKSMVL PKSMVLU DOSE DOSEUNI AT.recalc AE.recalc AEPCT.recalc
11      0     12  1.350   6.517  11300  NG/ML    560      ML   30      MG     6.328    10.046     33.48667
12     12     24 12.033  20.883   5720  NG/ML    650      ML   30      MG     3.718    10.046     33.48667
> y
  AMT.6.52 AMT.20.88 AE.6.52 AEPCT.6.52     AE   AEPCT1
1    6.328     3.718  10.046   33.48667 10.046 33.48667
> 
> sdeid <- 102652097 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE1, mct$DOSE1U)]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML") * 10^-6
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD PKATPDE PKCNCN PKCNCU PKSMVL PKSMVLU DOSE DOSEUNI AT.recalc AE.recalc AEPCT.recalc
77      0     12  0.567   11.85  32800  NG/ML   2420      ML  200      MG    79.376   107.114       53.557
78     12     24 14.467   23.75  20700  NG/ML   1340      ML  200      MG    27.738   107.114       53.557
> y
  AMT.11.85 AMT.23.75 AE.23.75 AEPCT.23.75      AE AEPCT1
3    79.376    27.738  107.114      53.557 107.114 53.557
> 
```
Next will be to review the tc111 results with this commit.

**Comment by tensfeldt** [comment]
For tc111, here mct$TIME=="ACTUAL",

For at least SDEID==2210432242 there are no "duplicate" times. However the same issue as was observed previously with tc018 occurs in that duplicate values are reported for the AE amounts for many of the time points. Also there are no AMT parameter values reported which is unexpected.

```r
> sdeid <- 2210432242 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE, mct$DOSEU)]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML")
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD_ORIG PKATPDE_ORIG  PKCNCN PKCNCU PKSMVL PKSMVLU    DOSE DOSEUNI   AT.recalc AE.recalc AEPCT.recalc
1       0     12       0.000       12.067 425.000 DPM/ML 1989.8      ML 1098776     DPM 845665.0000  896754.5     81.61395
2       0      0       0.000        0.000   0.000 DPM/ML  682.0      ML 1098776     DPM      0.0000  896754.5     81.61395
3      12     24      12.067       24.067  28.900 DPM/ML 1087.0      ML 1098776     DPM  31414.3000  896754.5     81.61395
4      24     48      24.067       48.100   3.080 DPM/ML 3568.2      ML 1098776     DPM  10990.0560  896754.5     81.61395
5      48     72      48.100       72.067   1.340 DPM/ML 3366.6      ML 1098776     DPM   4511.2440  896754.5     81.61395
6      72     96      72.067       96.067   0.710 DPM/ML 3289.0      ML 1098776     DPM   2335.1900  896754.5     81.61395
7      96    120      96.067      120.017   0.337 DPM/ML 3659.5      ML 1098776     DPM   1233.2515  896754.5     81.61395
8     120    144     120.017      144.033   0.112 DPM/ML 4302.9      ML 1098776     DPM    481.9248  896754.5     81.61395
9     144    168     144.033      168.017   0.030 DPM/ML 4116.9      ML 1098776     DPM    123.5070  896754.5     81.61395
10    168    192     168.017      192.017   0.000 DPM/ML 4267.1      ML 1098776     DPM      0.0000  896754.5     81.61395
> y
   AE.12.07  AE.24.07 AE.48.10 AE.72.07  AE.96.07 AE.120.02 AE.144.03 AE.168.02 AE.192.02 AEPCT.12.07 AEPCT.24.07 AEPCT.48.10 AEPCT.72.07 AEPCT.96.07 AEPCT.120.02 AEPCT.144.03
3 896754473 891761305 50965966 50965966 878794476 896754473  50965966  46096305 858990246    81.61395    81.15952    4.638431    4.638431     79.9794     81.61395     4.638431
  AEPCT.168.02 AEPCT.192.02        AE    AEPCT
3     4.195241     78.17701 896754473 81.61395
```

**Comment by jhhughes256** [comment]
With commit fbb93e0, checking the AET column numbers against the mct$ACTENDTIME shows no mismatches for tc018 and tc111. I have no comment on whether the numbers are correct though.

```r
> # SDEID with different column numbers than mct$ACTENDTIME values
> # tc018
> compare_colnums_endtimes()
[1] SDEID data 
<0 rows> (or 0-length row.names)
> # tc111 
> compare_colnums_endtimes()
[1] SDEID data 
<0 rows> (or 0-length row.names)
```

**Comment by tensfeldt** [comment]
With commit fbb93e0, the specific issues associated with this issue are resolved.

However a separate issue concerning conversion of units in DPMs to AMOUNTS in MG is highlighted below for context but will be submitted separately as well.

## For tc018 the incremental AEs are now correctly allocated to the endpoints of the associated intervals.

```r
> sdeid <- 43456658 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE1, mct$DOSE1U)]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML") * 10^-6
> x$ATPCT.recalc <- 100*x$AT.recalc/x$DOSE
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD PKATPDE PKCNCN PKCNCU PKSMVL PKSMVLU DOSE DOSEUNI AT.recalc ATPCT.recalc AE.recalc AEPCT.recalc
11      0     12  1.350   6.517  11300  NG/ML    560      ML   30      MG     6.328     21.09333    10.046     33.48667
12     12     24 12.033  20.883   5720  NG/ML    650      ML   30      MG     3.718     12.39333    10.046     33.48667
> y
  AMT.6.52 AMT.20.88 AE.6.52 AE.20.88 AEPCT.6.52 AEPCT.20.88     AE   AEPCT1
1    6.328     3.718   6.328   10.046   21.09333    33.48667 10.046 33.48667
> sdeid <- 102652097 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE1, mct$DOSE1U)]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML") * 10^-6
> x$ATPCT.recalc <- 100*x$AT.recalc/x$DOSE
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD PKATPDE PKCNCN PKCNCU PKSMVL PKSMVLU DOSE DOSEUNI AT.recalc ATPCT.recalc AE.recalc AEPCT.recalc
77      0     12  0.567   11.85  32800  NG/ML   2420      ML  200      MG    79.376       39.688   107.114       53.557
78     12     24 14.467   23.75  20700  NG/ML   1340      ML  200      MG    27.738       13.869   107.114       53.557
> y
  AMT.11.85 AMT.23.75 AE.11.85 AE.23.75 AEPCT.23.75 AEPCT.11.85      AE AEPCT1
3    79.376    27.738   79.376  107.114      53.557      39.688 107.114 53.557
```

## For tc111, "AT" for the incremental AMT values was added to the PARAMETERLIST

Here for SDEID==2210432242 and commit fbb93e0 the incremental AMTs are correctly reported, the cumulative AEs at the end of each sampling interval are correctly reported. 

```r
> sdeid <- 2210432242 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE, mct$DOSEU)]
> x <- x[order(x$PKATPD_ORIG, x$PKATPDE_ORIG),]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML")
> x$AE.cumsum <- cumsum(x$AT.recalc)
> x$ATPCT.recalc <- 100*x$AT.recalc/x$DOSE
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD_ORIG PKATPDE_ORIG  PKCNCN PKCNCU PKSMVL PKSMVLU    DOSE DOSEUNI   AT.recalc AE.cumsum ATPCT.recalc AE.recalc AEPCT.recalc
2       0      0       0.000        0.000   0.000 DPM/ML  682.0      ML 1098776     DPM      0.0000       0.0   0.00000000  896754.5     81.61395
1       0     12       0.000       12.067 425.000 DPM/ML 1989.8      ML 1098776     DPM 845665.0000  845665.0  76.96427661  896754.5     81.61395
3      12     24      12.067       24.067  28.900 DPM/ML 1087.0      ML 1098776     DPM  31414.3000  877079.3   2.85902677  896754.5     81.61395
4      24     48      24.067       48.100   3.080 DPM/ML 3568.2      ML 1098776     DPM  10990.0560  888069.4   1.00020896  896754.5     81.61395
5      48     72      48.100       72.067   1.340 DPM/ML 3366.6      ML 1098776     DPM   4511.2440  892580.6   0.41056994  896754.5     81.61395
6      72     96      72.067       96.067   0.710 DPM/ML 3289.0      ML 1098776     DPM   2335.1900  894915.8   0.21252648  896754.5     81.61395
7      96    120      96.067      120.017   0.337 DPM/ML 3659.5      ML 1098776     DPM   1233.2515  896149.0   0.11223866  896754.5     81.61395
8     120    144     120.017      144.033   0.112 DPM/ML 4302.9      ML 1098776     DPM    481.9248  896631.0   0.04386015  896754.5     81.61395
9     144    168     144.033      168.017   0.030 DPM/ML 4116.9      ML 1098776     DPM    123.5070  896754.5   0.01124042  896754.5     81.61395
10    168    192     168.017      192.017   0.000 DPM/ML 4267.1      ML 1098776     DPM      0.0000  896754.5   0.00000000  896754.5     81.61395
> y
  AMT.0.00 AMT.12.07 AMT.24.07 AMT.48.10 AMT.72.07 AMT.96.07 AMT.120.02 AMT.144.03 AMT.168.02 AMT.192.02 AE.0.00  AE.12.07  AE.24.07  AE.48.10  AE.72.07  AE.96.07 AE.120.02
3        0 845665000  31414300  10990056   4511244   2335190    1233252   481924.8     123507          0       0 845665000 877079300 888069356 892580600 894915790 896149042
  AE.144.03 AE.168.02 AE.192.02 AEPCT.0.00 AEPCT.12.07 AEPCT.24.07 AEPCT.48.10 AEPCT.72.07 AEPCT.96.07 AEPCT.120.02 AEPCT.144.03 AEPCT.168.02 AEPCT.192.02        AE    AEPCT
3 896630966 896754473 896754473          0    76.96428     79.8233    80.82351    81.23408    81.44661     81.55885     81.60271     81.61395     81.61395 896754473 81.61395 
```
## This behavior is confirmed for SDEID == 3021231022 

> sdeid <- 3021231022 
> x <- d[d$SDEID==sdeid, c(mct$NOMTIME, mct$NOMENDTIME, mct$ACTTIME, mct$ACTENDTIME, mct$CONC, mct$CONCU, mct$SAMPLEVOLUME, mct$SAMPLEVOLUMEU, mct$DOSE, mct$DOSEU)]
> x <- x[order(x$PKATPD_ORIG, x$PKATPDE_ORIG),]
> x$AT.recalc <- x[,mct$CONC] * x[,mct$SAMPLEVOLUME]/specific_gravity_adjustment(amt_units="ML")
> x$AE.cumsum <- cumsum(x$AT.recalc)
> x$ATPCT.recalc <- 100*x$AT.recalc/x$DOSE
> x$AE.recalc <- sum(x$AT.recalc)
> x$AEPCT.recalc <- 100*x$AE.recalc/x$DOSE
> y <- r[r$SDEID==sdeid,grep("(AMT|AE|AEPCT)(.[0-9]+?.[0-9]+?)*?", names(r), ignore.case=TRUE, perl=TRUE)]
> k <- unlist(lapply(y, FUN=function(x) { all(!is.na(x))}))
> y <- y[,names(k[k])]
> x
   PKPTMS PKPTME PKATPD_ORIG PKATPDE_ORIG   PKCNCN PKCNCU PKSMVL PKSMVLU    DOSE DOSEUNI    AT.recalc AE.cumsum ATPCT.recalc AE.recalc AEPCT.recalc
18      0      0       0.000        0.000   0.0000 DPM/ML  138.4      ML 1099202     DPM      0.00000       0.0  0.000000000  939400.7     85.46206
19      0     12       0.000       12.067 732.0000 DPM/ML 1220.6      ML 1099202     DPM 893479.20000  893479.2 81.284349919  939400.7     85.46206
20     12     24      12.067       24.067  49.4000 DPM/ML  660.2      ML 1099202     DPM  32613.88000  926093.1  2.967050642  939400.7     85.46206
21     24     48      24.067       48.083   9.5200 DPM/ML 1206.6      ML 1099202     DPM  11486.83200  937579.9  1.045015566  939400.7     85.46206
22     48     72      48.083       72.083   0.8160 DPM/ML 2123.6      ML 1099202     DPM   1732.85760  939312.8  0.157646875  939400.7     85.46206
23     72     96      72.083       96.067   0.0402 DPM/ML 2186.3      ML 1099202     DPM     87.88926  939400.7  0.007995733  939400.7     85.46206
24     96    120      96.067      120.017   0.0000 DPM/ML 1840.7      ML 1099202     DPM      0.00000  939400.7  0.000000000  939400.7     85.46206
25    120    144     120.017      144.033   0.0000 DPM/ML 2674.3      ML 1099202     DPM      0.00000  939400.7  0.000000000  939400.7     85.46206
26    144    168     144.033      168.017   0.0000 DPM/ML 2135.6      ML 1099202     DPM      0.00000  939400.7  0.000000000  939400.7     85.46206
27    168    192     168.017      192.017   0.0000 DPM/ML 2091.5      ML 1099202     DPM      0.00000  939400.7  0.000000000  939400.7     85.46206
> y
  AMT.0.00 AMT.12.07 AMT.24.07 AMT.48.08 AMT.72.08 AMT.96.07 AMT.120.02 AMT.144.03 AMT.168.02 AMT.192.02 AE.0.00  AE.12.07  AE.24.07  AE.48.08  AE.72.08  AE.96.07 AE.120.02
6        0 893479200  32613880  11486832   1732858  87889.26          0          0          0          0       0 893479200 926093080 937579912 939312770 939400659 939400659
  AE.144.03 AE.168.02 AE.192.02 AEPCT.0.00 AEPCT.12.07 AEPCT.24.07 AEPCT.48.08 AEPCT.72.08 AEPCT.96.07 AEPCT.120.02 AEPCT.144.03 AEPCT.168.02 AEPCT.192.02        AE    AEPCT
6 939400659 939400659 939400659          0    81.28435     84.2514    85.29642    85.45406    85.46206     85.46206     85.46206     85.46206     85.46206 939400659 85.46206
```

## UNITS conversion for DPM and MG appears incorrect

See the results above for SDEID==3021231022 and the detailed units STDOUT messages below.

For dose (in DPM) conversion to amounts in MG (through AMOUNTOUTPUTUNIT == "MG"), concentrations in DPM/ML and AMT and AE values are scaled to be 100 x the correct value (see above). The units_conversion.R code sets the DPM to MG scaling factor to 1 which is expected.

```r
> results_list <- run_computation(data=d, map=mct, flag=flags, parameterset=parameterset)
unit_conversion : Unit Class 1 (Time) time_col:  TMAXRATE MIDPTLAST TLAG MIDPT1 MIDPT2 MIDPT3 MIDPT4 MIDPT5 MIDPT6 MIDPT7 MIDPT8 MIDPT9 MIDPT10 MIDPT11 MIDPT12  parameters are scaled from  HR  to  HR  via scaling factor:  1 
unit_conversion : Unit Class 2 (Amount) amount_col:  AMT.0.00 AMT.11.05 AMT.11.07 AMT.11.08 AMT.12.03 AMT.12.05 AMT.12.07 AMT.23.05 AMT.24.05 AMT.24.07 AMT.24.08 AMT.47.05 AMT.47.07 AMT.48.05 AMT.48.07 AMT.48.08 AMT.48.10 AMT.48.15 AMT.71.05 AMT.71.07 AMT.71.08 AMT.72.03 AMT.72.05 AMT.72.07 AMT.72.08 AMT.95.03 AMT.95.05 AMT.95.07 AMT.96.07 AMT.96.08 AMT.96.17 AMT.119.02 AMT.119.03 AMT.120.02 AMT.143.05 AMT.143.07 AMT.143.10 AMT.144.00 AMT.144.02 AMT.144.03 AMT.144.10 AMT.168.02 AMT.168.05 AMT.168.18 AMT.192.02 AMT.192.10 AMT.216.07 AMT.240.02 AE.0.00 AE.11.05 AE.11.07 AE.11.08 AE.12.03 AE.12.05 AE.12.07 AE.23.05 AE.24.05 AE.24.07 AE.24.08 AE.47.05 AE.47.07 AE.48.05 AE.48.07 AE.48.08 AE.48.10 AE.48.15 AE.71.05 AE.71.07 AE.71.08 AE.72.03 AE.72.05 AE.72.07 AE.72.08 AE.95.03 AE.95.05 AE.95.07 AE.96.07 AE.96.08 AE.96.17 AE.119.02 AE.119.03 AE.120.02 AE.143.05 AE.143.07 AE.143.10 AE.144.00 AE.144.02 AE.144.03 AE.144.10 AE.168.02 AE.168.05 AE.168.18 AE.192.02 AE.192.10 AE.216.07 AE.240.02 AE  parameters are scaled from  DPM  to  MG  via scaling factor:  1000 
unit_conversion : Unit Class 3 (Dose) dose_col:  DOSE  parameters are scaled from  DPM  to  MG  via scaling factor:  1000 
unit_conversion : Unit Class 4 (Volume) volume_col:  VOLSUM  parameters are scaled from  ML  to  ML  via scaling factor:  1 
unit_conversion : Unit Class 11: ([Volume.Amount]/Volume) aurc_col:  AURCALL AURCLAST AURC1 AURC2 AURC3 AURC4 AURC5 AURC6 AURC7 AURC8 AURC9 AURC10 AURC11 AURCINFO AURCINFP  parameters are scaled from  ML.DPM/ML  to  ML.UG/ML  via scaling factor:  1e+06 
unit_conversion : Unit Class 13: (Amount/Time) rate_col:  MAXRATE RATELAST RATE1 RATE2 RATE3 RATE4 RATE5 RATE6 RATE7 RATE8 RATE9 RATE10 RATE11 RATE12  parameters are scaled from  DPM/HR  to  UG/HR  via scaling factor:  1e+06 
There were 24 warnings (use warnings() to see them)
```

Based upon this testing and opening a new issue for the DPM conversion issue, this issue can be closed.
